test-plans/Dockerfile

```diff
@@ -0,0 +1,19 @@
+FROM rust:1.65-bullseye as builder
+WORKDIR /usr/src/libp2p
+
+# TODO fix this, it breaks reproducibility
+RUN apt-get update && apt-get install -y cmake protobuf-compiler
+
+COPY ./ ./
```

Comment:
Copying everything will invalidate this layer every time you modify anything. You need to copy files in two steps:

1. Copy `Cargo.toml` + `Cargo.lock` + `src/main.rs`
2. Run `cargo build`
3. Copy `src/lib.rs` and `src/bin/ping.rs`

Because the `main.rs` is a dummy file, the docker layer created in step (2) does not get invalidated when you work on `src/bin/ping.rs` or `src/lib.rs`.

Reply:
See https://github.com/libp2p/test-plans/blob/master/ping/rust/Dockerfile as an example @jxs.

Reply:
yeah I was researching the Docker in Docker option raised by Thomas on https://github.com/libp2p/rust-libp2p/pull/3331#discussion_r1073532935 cause that was the reason for having this. As we are depending on `rust-libp2p` from the parent dir, it needs to be copied along `transports` and `protocols`, I take it you prefer to copy all that instead of having this right?

Reply:
I think we are better off with:

1. making this crate part of the main workspace
2. adding our usual caching layer in github actions (with `save-if: <master branch>`)
3. building the binary outside of docker
4. have the dockerfile copy in the binary from the root target directory

The cache is reusable across all our jobs (master, PRs and tags) so we get maximum cache hits and the configuration is very simple. The only downside is that MacOS and Windows users cannot build this docker image locally. I am okay with that:

- We have no MacOS or Windows users among the core maintainers
- The Docker container isn't a strict requirement for the tests, the binary also works standalone